llvm/lib/Target/AMDGPU/SIISelLowering.cpp

```diff
@@ -6899,7 +6899,17 @@ SDValue SITargetLowering::lowerFP_ROUND(SDValue Op, SelectionDAG &DAG) const {
     if (Op.getOpcode() != ISD::FP_ROUND)
       return Op;
 
-    SDValue FpToFp16 = DAG.getNode(ISD::FP_TO_FP16, DL, MVT::i32, Src);
+    if (!Subtarget->has16BitInsts()) {
+      SDValue FpToFp16 = DAG.getNode(ISD::FP_TO_FP16, DL, MVT::i32, Src);
+      SDValue Trunc = DAG.getNode(ISD::TRUNCATE, DL, MVT::i16, FpToFp16);
+      return DAG.getNode(ISD::BITCAST, DL, MVT::f16, Trunc);
+    }
```

Comment:
I'm confused by this. FP_TO_FP16 does the right thing for an f64 source? Should the unsafe path be before this?

Reply:
Before this patch, FP_TO_FP16 is used for all targets for FP_ROUND f64 source so I think it suppose to do the right thing. However, I took a quick look at lit tests and we seems don't have any lit test that covers FP_ROUND f64 source in targets that do not support f16 types.

So I would say the FP_TO_FP16 here at least keeps the same routine for targets that do not support f16 types.

Reply:
Hi Matt. Do you have any comment on this? Thanks!

Reply:
> we seems don't have any lit test that covers FP_ROUND f64 source in the old targets that do not support f16 types.

This is covered, fptrunc.ll has it. fptrunc.f16.ll should also cover it, but that one brokenly only tests with -enable-unsafe-fp-math 

It seems to work because we have explicit custom lowering f64 source. Seems like an unnecessary complication over the default expansion

Reply:
I see. I just realized SI is gfx6. Will create another patch to address these